tools/expotools/src/Git.ts

```diff
@@ -0,0 +1,300 @@
+import path from 'path';
+
+import { spawnAsync, SpawnResult, SpawnOptions } from './Utils';
+import { EXPO_DIR } from './Constants';
+
+export type GitPullOptions = {
+  rebase?: boolean;
+};
+
+export type GitPushOptions = {
+  track?: string;
+};
+
+export type GitLogOptions = {
+  fromCommit?: string;
+  toCommit?: string;
+  paths?: string[];
+};
+
+export type GitLog = {
+  hash: string;
+  parent: string;
+  title: string;
+  authorName: string;
+  committerRelativeDate: string;
+};
+
+export type GitFileLog = {
+  path: string;
+  relativePath: string;
+  status: GitFileStatus;
+};
+
+export enum GitFileStatus {
+  M = 'modified',
+  C = 'copy',
+  R = 'rename',
+  A = 'added',
+  D = 'deleted',
+  U = 'unmerged',
+}
+
+export type GitBranchesStats = {
+  ahead: number;
+  behind: number;
+};
+
+export type GitCommitOptions = {
+  title: string;
+  body?: string;
+};
+
+/**
+ * Helper class that stores the directory inside the repository so we don't have to pass it many times.
+ * This directory path doesn't have to be the repo's root path,
+ * it's just like current working directory for all other commands.
+ */
+export class GitDirectory {
+  readonly path: string;
+  readonly Directory = GitDirectory;
+
+  constructor(path) {
+    this.path = path;
+  }
+
+  /**
+   * Generic command used by other methods. Spawns `git` process at instance's repository path.
+   */
+  async runAsync(args: string[], options: SpawnOptions = {}): Promise<SpawnResult> {
+    return await spawnAsync('git', args, {
+      cwd: this.path,
+      ...options,
+    });
+  }
+
+  /**
+   * Same as `runAsync` but returns boolean value whether the process succeeded or not.
+   */
+  async tryAsync(args: string[], options: SpawnOptions = {}): Promise<boolean> {
+    try {
+      await this.runAsync(args, options);
+      return true;
+    } catch (error) {
+      return false;
+    }
+  }
+
+  /**
+   * Returns repository's branch name that you're checked out on.
+   */
+  async getCurrentBranchNameAsync(): Promise<string> {
+    const { stdout } = await this.runAsync(['rev-parse', '--abbrev-ref', 'HEAD']);
+    return stdout.replace(/\n+$/, '');
+  }
+
+  /**
+   * Returns name of remote branch that the current local branch is tracking.
+   */
+  async getTrackingBranchNameAsync(): Promise<string> {
+    const { stdout } = await this.runAsync([
+      'rev-parse',
+      '--abbrev-ref',
+      '--symbolic-full-name',
+      '@{u}',
+    ]);
+    return stdout.trim();
+  }
+
+  /**
+   * Tries to deduce the SDK version from branch name. Returns null if the branch name is not a release branch.
+   */
+  async getSDKVersionFromBranchNameAsync(): Promise<string | null> {
+    const currentBranch = await this.getCurrentBranchNameAsync();
+    const match = currentBranch.match(/\bsdk-(\d+)$/);
+
+    if (match) {
+      const sdkMajorNumber = match[1];
+      return `${sdkMajorNumber}.0.0`;
```

Comment:
We need to keep in mind that we'll probably change from `SDK-X.X.X` to `SDK-X` 😉

Reply:
Yeah, but that's separate thing and we shouldn't change this until we migrate to the new style. The body of this method was just moved from another file without modifications.